### PR TITLE
move from pio to idf toolchain

### PR DESCRIPTION
**Pull request by nopeslide**
regarding #49 

- [x] add idf docker image
- [x] add idf docker container
- [x] rebase other docker images on idf
  - [x] tested qemu container
  - [x] tested vscode container
  - [x] remove pio docker images/container
- [x] restructure repo
- [x] modify main Makefile to use idf docker
  - [x] tested build
  - [x] tested monitor
  - [x] tested flash
  - [x] tested clean
  - [x] tested distclean
  - [x] add menuconfig target
- [x] binary works
  - [x] compiles
  - [x] ulp works
  - [x] binary does not crash
- [x] documentation

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -49,13 +48,13 @@ TARGET += flash
 HELP_flash = flashes project to esp. Use DEV=path to provide  path to device or use make dev
 flash: | check-flash
 	@make --no-print-directory -C docker \
-		pio \
-		EXEC="sudo chgrp developer $(DEV); pio run --upload-port '$(DEV)' $(PIO_ENV_OPTION) -t upload" DEV="$(DEV)"
+		idf.py \
```
Comment (jhonnyam) [inline Makefile]: Should be just `idf`

File: Makefile [hunk 2]
```diff
@@ -65,17 +64,12 @@ HELP_monitor = connects to esp32 via serial. Use DEV=path to provide path to dev
 monitor: | check-monitor
 	@make --no-print-directory -C docker \
 		pio \
```
Comment (jhonnyam) [inline Makefile]: should be `idf`

**Review by nopeslide — COMMENTED** [review]
seems reviewable.
a little bit more documentation needs to be added

**Review by jhonnyam — COMMENTED** [review]
Looks good

File: components/template/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,21 @@
+idf_component_register(
+  SRC_DIRS
+    src
+	INCLUDE_DIRS
```
Comment (jhonnyam) [inline components/template/CMakeLists.txt]: Bad indentation

**Review by jhonnyam — CHANGES_REQUESTED** [review]
Looks good

**Review by jhonnyam — APPROVED** [review]

**Pull request merged**